### PR TITLE
Allow configurable `-l select` options in PBSPro provider

### LINKED ISSUE #2220: PBSPro scheduler options ignored
**Describe the bug**
The PBSPro provider template adds `${scheduler_options}` to the bottom of the list of #PBS commands. However, PBSPro seems to only consider the first #PBS option and ignore any later competing ones. This means specifying a new select option with gpus is superseded by the default one and ignored. We can resolve this by moving the user defined scheduler options to the top of the #PBS list in the template.

**To Reproduce**
Use the PBSPro provider (e.g., at ALCF's edge testbed for Polaris) and try to specify a new select option:`-l select=1:ncpus=32:ngpus=1`. These options are ignored in favor of the default specified above in the template.

**Expected behavior**
I would like scheduler_options to be put first so they can be used to enter specific options.

**Environment**
 - Parsl 1.2.0

**Distributed Environment**
- Where are you running the Parsl script from ? ALCF
- Where do you need the workers to run ? Compute nodes

### PR DESCRIPTION
**Pull request by ryanchard**
# Description

Moves the PBSPro scheduler option to the top of the template. 

Fixes #2220 

## Type of change

- Bug fix (non-breaking change that fixes an issue)

### CODE REVIEW

**Comment by benclifford** [comment]
Does this override this line entirely:

> #PBS -l select=${nodes_per_block}:ncpus=${ncpus}

?

and if so, maybe there should be a parameter to allow additional selects to be added on this line...

**Review by yadudoc — COMMENTED** [review]
This bothers me endlessly. Some implementations/installations use different ordering, and we might be better served by placing the options both at the top and bottom assuming one would be ignored. Alternatively we could do something like `scheduler_opt_placement = {'top', 'bottom'}`. 

@benclifford what do you think?

**Comment by yadudoc** [comment]
I'm okay with going with this now, since the only supported/tested target site is polaris.

**Comment by benclifford** [comment]
@yadudoc is it `-l select` overrides that you are doing or other PBS overrides?

If pre and postfix behaviour genuinely is different per site, then a scheduler pre options and scheduler post options which need to be configured according to how your site wants things is one way to do it - and, the `scheduler_options` field as a single field does not make sense, as something that magically will be pasted in the "right" place.

**Comment by benclifford** [comment]
of the fields that the pbs provider sets, `-l select` is the only one i'd expect someone be expecting to override with their own behaviour, right?

**Comment by ryanchard** [comment]
Good question on whether that is the only field people would need to override. I originally assumed there would be others, but perhaps not. Anything else you'd want to customize probably won't need to overwrite an existing, non-configurable field in the template. In my case I just I want to specify GPUs with an addition of `:ngpus=2` to the end of the select -l statement. An option to extend the existing select field would also work.

**Comment by benclifford** [comment]
@ryanchard what i think should happen here is that the pbs provider gets a new option for setting/overriding the `-l select` field, rather than hoping (unreliably) that placing a second `-l select` elsewhere in the job definition will work.

File: parsl/providers/pbspro/pbspro.py [hunk 1]
```diff
@@ -119,6 +124,12 @@ def submit(self, command, tasks_per_node, job_name="parsl"):
         job_config["worker_init"] = self.worker_init
         job_config["user_script"] = command
 
+        # Add a colon to select_options if one isn't included
+        if self.select_options and not self.select_options.startswith(":"):
```
Comment (benclifford) [inline parsl/providers/pbspro/pbspro.py:128]: if I put a `:` on the front, this silently ignores the options rather than erroring?

**Comment by benclifford** [comment]
btw 

> I'm okay with going with this now, since the only supported/tested target site is polaris.

@yadudoc i have other users using pbspro that are not polaris

File: parsl/providers/pbspro/pbspro.py (hunk 1 above)
Comment (benclifford) [reply parsl/providers/pbspro/pbspro.py:128]: oh ignore this i misunderstood

**Pull request merged**